src/daemon.rs

```diff
@@ -138,3 +145,305 @@ impl fmt::Display for NotifyState {
+///
+/// https://www.freedesktop.org/software/systemd/man/systemd.socket.html
+#[derive(Debug, Clone)]
+pub enum SocketType {
+    /// A FIFO named pipe (see man 7 fifo)
+    Fifo(RawFd),
+    /// A special file, such as character device nodes or special files in
+    /// /proc and /sys
+    Special(RawFd),
+    /// A PF_INET socket, such as UDP/TCP sockets
+    Inet(RawFd),
+    /// A PF_UNIX socket (see man 7 unix)
+    Unix(RawFd),
+    /// A POSIX message queue (see man 7 mq_overview)
+    Mq(RawFd),
+}
+
+impl IsType for SocketType {
+    fn is_fifo(&self) -> bool {
+        match self {
+            SocketType::Fifo(_) => true,
+            _ => false,
+        }
+    }
+
+    fn is_special(&self) -> bool {
+        match self {
+            SocketType::Special(_) => true,
+            _ => false,
+        }
+    }
+
+    fn is_unix(&self) -> bool {
+        match self {
+            SocketType::Unix(_) => true,
+            _ => false,
+        }
+    }
+
+    fn is_inet(&self) -> bool {
+        match self {
+            SocketType::Inet(_) => true,
+            _ => false,
+        }
+    }
+
+    fn is_mq(&self) -> bool {
+        match self {
+            SocketType::Mq(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// Check for file descriptors passed by systemd
+///
+/// Invoked by socket activated daemons to check for file descriptors needed by the service.
+/// If unset_env is true, the environment variables used by systemd will be cleared.
+pub fn sd_listen_fds(unset_env: bool) -> Result<Vec<SocketType>> {
+    let pid = env::var("LISTEN_PID")?;
+    let fds = env::var("LISTEN_FDS")?;
+    if unset_env {
+        env::remove_var("LISTEN_PID");
+        env::remove_var("LISTEN_FDS");
+        env::remove_var("LISTEN_FDNAMES");
+    }
+
+    let pid = pid.parse::<u32>()?;
+    let fds = fds.parse::<i32>()?;
+
+    if process::id() != pid {
+        return Err("Pid mismatch".into());
+    }
+
+    let vec = socks_from_fds(fds);
+    Ok(vec)
+}
+
+/// Check for file descriptors passed by systemd
+///
+/// Like sd_listen_fds, but will also return a Vec of names associated with each file
+/// descriptor.
+pub fn sd_listen_fds_with_names(unset_env: bool) -> Result<Vec<(SocketType, String)>> {
+    let pid = env::var("LISTEN_PID")?;
+    let fds = env::var("LISTEN_FDS")?;
+    let names = env::var("LISTEN_FDNAMES")?;
+
+    if unset_env {
+        env::remove_var("LISTEN_PID");
+        env::remove_var("LISTEN_FDS");
+        env::remove_var("LISTEN_FDNAMES");
+    }
+
+    let pid = pid.parse::<u32>()?;
+    let fds = fds.parse::<i32>()?;
+
+    if process::id() != pid {
+        return Err("Pid mismatch".into());
+    }
+
+    let names: Vec<String> = names.split(":").map(String::from).collect();
+    let vec = socks_from_fds(fds);
+    let out = vec.into_iter().zip(names.into_iter()).collect();
+
+    Ok(out)
+}
+
+fn socks_from_fds(fds: RawFd) -> Vec<SocketType> {
```

Comment:
I think the type in `fds: RawFd` looks strange as you are doing a `+` on it later on. Isn't this just the length of the input array? Is there some way we can make this code less tricky to follow?

Reply:
Good point, fixed!